src/main/java/org/wildfly/security/auth/client/CredentialStoreReference.java

```diff
@@ -0,0 +1,134 @@
+ * @author <a href="mailto:pskopek@redhat.com">Peter Skopek</a>
+ */
+public final class CredentialStoreReference {
+    private final String store;
+    private final String alias;
+    private final Class<? extends Credential> type;
+    private final char[] clearText;
+
+    /**
+     * Creates credential store reference instance using following set of params.
+     *
+     * @param store the name of store to obtain the credential
+     * @param alias the alias of {@link Credential} withing the (credential) store
+     * @param typeName the credential type of wanted {@link Credential}
+     */
+    public CredentialStoreReference(String store, String alias, String typeName) {
+        this(store, alias, typeName, null, null);
+    }
+
+    /**
+     * Creates credential store reference instance using following set of params.
+     *
+     * @param store the name of store to obtain the credential
+     * @param alias the alias of {@link Credential} withing the (credential) store
+     */
+    public CredentialStoreReference(String store, String alias) {
+        this(store, alias, null, null, null);
+    }
+
+    /**
+     * Creates credential store reference instance using following set of params.
+     *
+     * @param store the name of store to obtain the credential
+     * @param alias the alias of {@link Credential} withing the (credential) store
+     * @param type the credential type of wanted {@link Credential}
+     */
+    public CredentialStoreReference(String store, String alias, Class<? extends Credential> type) {
+        this(store, alias, null, type, null);
+    }
+
+    /**
+     * Creates reference to simple clear-text password.
+     * @param clearText password in clear text form
+     */
+    public CredentialStoreReference(char[] clearText) {
+        this(null, null, null, null, clearText.clone());
+    }
+
+    private CredentialStoreReference(String store, String alias, String typeName, Class<? extends Credential> type, char[] clearText) {
+        this.store = store;
+        this.alias = alias;
+        if (type != null) {
+            this.type = type;
+        } else {
+            this.type = getCredentialType(typeName);
+        }
+        this.clearText = clearText;
+    }
+
+    /**
+     * Gets the credential store name
+     * @return store name
+     */
+    public String getStore() {
+        return store;
+    }
+
+    /**
+     * Gets the alias
+     * @return alias
+     */
+    public String getAlias() {
+        return alias;
+    }
+
+    /**
+     * Gets the credential type
+     * @return credential type
+     */
+    public Class<? extends Credential> getType() {
+        return type;
+    }
+
+    /**
+     * Gets the password in clear text form
+     * @return password
+     */
+    public char[] getClearText() {
+        return clearText;
+    }
+
+    private static <C extends Credential> Class<C> getCredentialType(String typeName) {
```

Comment:
This kind of method is not OK because it does an unsafe cast of the returned class object.  Normally you want to hand in a Class that has the known bound, and use asSubclass() to narrow it safely..